Compound commands

nclsbsta opened the pull request.

Commit:
fix tokenization bug when calling history line with heredoc in it, fix strcpy and binding

Commit:
Add subshells parsing, add basic subshells execution

Commit:
fix bug when executing listed commands

Commit:
add eval_group_command, slightly refactor execution

Commit:
add { parsing and temporary incorrect reserved words handling

Commit:
add t_group_cmd, cleanup

Commit:
add g_lexer.expect_reserv_word, move redir_list from t_command to t_group_cmd,
move first call to get_next_token() from main_loop to parse_complete_command.

Commit:
use clang to compile libft

Commit:
avoid extra fork for compound commands with only one command

Commit:
fix subshell problems due to previous commit

Commit:
fix bugs due to 2 previous commits, simplify subshells handling

Commit:
fix nested subshell parsing problem

Commit:
fix missing fork in subshell listed commands

Commit:
move something from somewhere to somewhere else

Commit:
fix bug in nested subshell parsing optimisation

Commit:
add bracket nest limit, cleanup files

Commit:
add tests for group commands, expect reserved words after pipes and add mandatory_reserv_word() function

Commit:
prevent mandatory_reserv_word from segv

Commit:
add basic async_lists handling in group commands, fix leaks in group commands parsing, fix readline history bug

fratajczak committed:
Ignore ODR violations like gcc

Commit:
fix leak in command_del

Commit:
norme, attempt #1

Commit:
Merge branch 'compound_commands' of https://github.com/42shell/42sh into compound_commands

Commit:
norme, attempt #2

fratajczak committed:
Use gcc in travis for coverage

Commit:
fix segv when "{ ls ; ls -z } | 2>& cat -e, line continuation, ^C", add comment for parse_sperator functions

Commit:
Merge branch 'compound_commands' of https://github.com/42shell/42sh into compound_commands

Commit:
fix parse error bug due to wrong error handling in parse_and_or and parse_pipeline, add test for this particular case

Commit:
fix previous fix: check parse errors in parse_command, parse_compound_command, parse_simple_command

Commit:
fix syntax error

Commit:
remove stdio.h

fratajczak commented on include/lexer.h:
```diff
@@ -22,8 +22,8 @@
 **         '<<'   '>>'    '<&'     '>&'      '<>'         <<-        '&'
-**		  SEMI    PIPE    LESS   GREAT
-**		   ';'	   '|'     '<'    '>'
+**		  SEMI    PIPE    LESS   GREAT	OBRACKET	CBRACKET
```
le commentaire est faux

nclsbsta replied:
fixed

fratajczak commented on include/lexer.h:
```diff
@@ -46,6 +46,10 @@ enum				e_token_type
 	AND_IF,
 	OR_IF,
 	BANG,
+	LBRACKET,
```
c'est lparen

nclsbsta replied:
je laisse comme ça j'ai la flemme, tu feras un sed si vraiment ça te gêne

fratajczak commented on include/lexer.h:
```diff
@@ -62,50 +66,53 @@ enum				e_quote_st
+	char				nl_found;
+	char				end_of_input;
+	bool				expect_reserv_word;
+	size_t				i;
```
l'interieur des structs doit pas etre aligne juste le s_lexer t_lexer

nclsbsta replied:
c'est moi qui décide de ce qui doit être aligné ou non

fratajczak replied:
c'est juste que ca rajoute des changements pour rien en soi osef de l'alignement faut juste que ca soit a la norme

nclsbsta replied:
si c'est pas aligné c'est pas à ma norme

fratajczak commented on include/shell.h:
```diff
@@ -60,39 +64,43 @@ typedef struct			s_var
-** -error handling
-**		-^C ^D
-**		-last exit status, close fds... assert everything is perfect
+** todo:
```
pourquoi tu met des todo dans les .h serieux fait une issue

nclsbsta replied:
c'est plus pratique d'avoir ça dans le .h pour moi, que sur mon navigateur, personne te demande de le lire

fratajczak replied:
ca fait des kb en plus sur mon disque dur

fratajczak commented on libft/src/readline/binding.c:
```diff
@@ -14,12 +14,20 @@
+	char	*str;
+
+	if ((str = tgetstr("kD", NULL)))
+		strcpy((char *)&g_rl_keys[K_DEL], str);
```
ft

fratajczak commented on libft/src/readline/readline.c:
```diff
@@ -54,7 +54,7 @@ static int	rl_init(void)
 		g_rl_is_init = true;
 	}
 	tcsetattr(STDIN_FILENO, TCSADRAIN, &g_rl_term);
-	tputs(g_rl_caps[C_KS], 1, ft_putc);
```
met 1

fratajczak commented on libft/src/readline/term_cursor.c:
```diff
@@ -16,10 +16,10 @@ void	movcright(void)
 {
 	if (g_rl_posx == g_rl_sizex - 1)
 	{
-		tputs(tgoto(g_rl_caps[C_CH], 0, 0), 1, ft_putc);
```
1

fratajczak commented on src/execution/execute.c:
```diff
@@ -80,3 +80,49 @@ int			exec_simple_command(t_simple_cmd *simple)
+	t_process	*process;
+	int			fd;
+
+	fd = open("/dev/null", O_RDONLY);
```
il faut le mettre a dev null que si c'est en non interactif
mais je sais pas si ca change qqchose

nclsbsta replied:
fixed

fratajczak commented on src/execution/redirect.c:
```diff
@@ -77,13 +77,13 @@ static int	get_redirected_fd(t_redir *redir)
 	int		ret;
 
-	redir = command->redirs;
+	redir = redir_list;
```
lol il sert a quoi redir du coup

nclsbsta replied:
A générer des commentaires inutiles apparemment

Commit:
fix newline replacement bug in parse_newline_list

fratajczak commented on src/parser/complete_command.c:
```diff
@@ -0,0 +1,53 @@
+
+/*
+** -incorrect, complete_commands must be separated by newline_list,
+** commands separated by '\0' in batch mode will work.
```
met parse_newline_list

nclsbsta replied:
...

fratajczak commented on src/parser/simple_command.c:
```diff
@@ -84,30 +84,26 @@ static int		set_tokens(t_simple_cmd *simple)
-			&& !is_redir(g_parser.token)))
+	|| g_parser.token->type == NEWLINE)
+		return (NULL);
+	else if (g_parser.token->type != WORD
```
pourquoi

nclsbsta replied:
pourquoi pas

fratajczak commented on src/parser/utils.c:
```diff
@@ -0,0 +1,51 @@
+
+#include "shell.h"
+
+int			mandatory_reserv_word(int expect_type)
```
no consistency !!!!!

nclsbsta replied:
changed for get_required_reserv_word(int expect_type)

Commit:
fix handling of async lists when job control is disabled, + a few modifications related to your inconsistent comments

Commit:
change tputs padding from 20 to 1

Commit:
norme, attempt #1

fratajczak committed:
Merge master intro compound_commands

fratajczak committed:
Pop dparens at the 2nd paren in set_bracket_status

Because if we pop them at the first one, brack_stack is empty
and operators are recognized, which means the second paren is tokenized
as an operator

fratajczak committed:
Merge branch 'master' into compound_commands

Commit:
fix bg subshells bug, temp fix of lexer for non newline-terminated lines

Commit:
add temp wait_for_job in launch_process to avoid problems in implicit subshells "ls | cat &", update jobs display for group commands

Commit:
fix some job control bugs, add global bg variable, add a few tests for job control, norme jobs display

Commit:
make jobs work properly in batch mode, add some tests

Commit:
fix problem with files without ending newline almost properly, add test for it

Commit:
Merge branch 'compound_commands' of https://github.com/42shell/42sh into compound_commands

Commit:
add handling of EOF while looking for matching quote/bracket

Commit:
add parse error when UNEXPECTED EOF while looking for matching quote

Commit:
fix corrupted test, remove file
